File: Difficulty: Basic/Preorder Traversal/preorder-traversal.cpp
/*
class Node
{
    int data;
    Node* left;
    Node* right;

    Node(int x){
        data = x;
        left = right = NULL;
    }
};
*/

class Solution {
  public:
    void solve(Node* root, vector<int>& ans) {
        // root left right
        
        if(root==nullptr) return;
        ans.push_back(root->data);
        solve(root->left, ans);
        solve(root->right, ans);
    }
    vector<int> preOrder(Node* root) {
        // code here
        vector<int> ans;
        solve(root,ans);
        return ans;
    }
};

File: Difficulty: Basic/Preorder Traversal/preorder-traversal.java
/*
class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    public void solve(Node root, List<Integer> list) {
        if(root==null) return;
        list.add(root.data);
        solve(root.left,list);
        solve(root.right,list);
    }
    public ArrayList<Integer> preOrder(Node root) {
        //  code here
        
        List<Integer> list = new ArrayList<>();
        
        solve(root,list);
        
        return (ArrayList<Integer>)list;
        
    }
}